### PR TITLE
Enable mapping of source folders to target folders

### PR DESCRIPTION
**Pull request by mathieuvanloon**
Allows mapping entries of the form { "glob" : "targetdir" }. For example: 

```
    "sources": {
      "jquery-ui": {
        "mapping": [
          { "../bowerComponents/jquery-ui/jquery-ui.min.js": "jquery-ui.min.js" },
          { "../bowerComponents/jquery-ui/themes/smoothness/**": "smoothness" },
          { "../bowerComponents/jquery-ui/themes/base/**": "base" },
.. snip ..
```

### CODE REVIEW

**Comment by Biromain** [comment]
Any news?

**Comment by TachikomaGT** [comment]
Please, accept this PR. Without this functionality importing (for example) of twitter's bootstap less files becomes hell.

**Comment by rquadling** [comment]
Hi.

Can you rebase your changes off master and I'll get them merged.

Thank you for your efforts.

**Comment by rquadling** [comment]
BTW, I'm the new project owner, so hopefully things will move forward now.

**Comment by rquadling** [comment]
Can you also add some tests as I need to know that this works as intended and doesn't break any existing functionality.

**Comment by rquadling** [comment]
Can you show me a good example I can incorporate as a test please?

As soon as I've got an understanding of the intent, I can make sure the test works and then get things pulled and released.

And rebase please as things have moved on a little since this PR was made.

**Comment by rquadling** [comment]
Pinging @mathieuvanloon, @Biromain, @TachikomaGT

I need your help to build a test for this please. Once tested, I can merge. Thanks.

**Comment by rquadling** [comment]
Still would like to get this tested and merged, but please can you rebase on to the current master.

**Comment by mathieuvanloon** [comment]
Hi Richard,

sorry i completely missed your earlier requests. I'll take a look at it asap

2017-01-26 12:57 GMT+01:00 Richard Quadling <notifications@github.com>:

**Comment by rquadling** [comment]
Thank you!!!

File: lib/installer.js [hunk 1]
```diff
@@ -181,12 +186,39 @@ exports.installDependency = function installDependency(deps, pakcfg, cfg, paths,
...
+						}
+					} else {
+						async.each(files, function(f, callback) {
+							installFile(f, pakcfg, paths, src, dest, silent, options, callback);
+						}, callback);
+					}
+				});
+
+			} else {
+	            installFile(dep, pakcfg, paths, dep, '', silent, options, callback);
```
Comment (rquadling) [inline lib/installer.js]: Watch out for whitespace issues.

**Comment by rquadling** [comment]
@mathieuvanloon Hi. Any chance you could rebase please?

**Comment by Amenel** [comment]
I'd like to chime in on this. On our current project, we also need to import components (such as ckeditor) with files dispatched into folders and subfolders. Things become nightmarish because each single file from those folders (e.g. ckeditor skins) has to be mapped manually. Add to this the fact that components do not follow a fixed set of conventions: some have a "dist" folder, others don't, some put files in distinct folders, others don't.

This "problem" (apparently) already has a solution to be seen by all, except that the author no longer has time to work on this small body of code. Maybe that somebody else could simply reauthor this contribution and the maintainer can perform a release? I believe it would help numerous teams. I know it would ours.

**Comment by rquadling** [comment]
As the maintainer of this package, I am happy to work with people making PRs.

For this request, I'd like some tests. As the maintainer, I don't want to be making assumptions about how something works. As things stand, this PR is unmergable due to conflicts, and has been that way for some time.

Without tests, documentation and a mergeable PR, I'm happy to leave this unmerged until those conditions are met.

So, @Amenel , do you want to take the branch, clean it up, add appropriate tests and documentation? If so, I'll be VERY happy to merge a PR!

**Comment by mathieuvanloon** [comment]
Hi! Sorry I didn't act yet. It should not be a problem to prepare a proper pull request. However I am quite unfamiliar with git and I need some advice on how to proceed. What I'd like to do is just overwrite this branch with the latest state from rquadling and then manually reapply the changes to lib/installer.js. I think I just merged everything from rquadling except installer.js, but I am not entirely sure :) So what's the best way to go about this? thanks for your help!

**Comment by rquadling** [comment]
What you are asking about is a process called rebasing. This is where your changes are replayed onto the tip of a branch. I use Atlassian's SourceTree (Windows and MacOS clients). My colleagues use SmartGIT. Both very good products. From there you can set up 2 two remotes (this repo and your repo) and manage the merges/rebases with just a few clicks. Pushing the changes to your repo and making a pull request are also all within the UI.

**Comment by rquadling** [comment]
I've added a test and updated the documentation.

Also some white space changes.

**Pull request merged**